src/Workspaces/Core/Portable/Workspace/Solution/TextDocumentStates.cs

```diff
@@ -0,0 +1,226 @@
+                   infos.ToImmutableSortedDictionary(info => info.Id, stateConstructor, DocumentIdComparer.Instance))
+        {
+        }
+
+        public int Count
+            => _map.Count;
+
+        public bool IsEmpty
+            => Count == 0;
+
+        public bool Contains(DocumentId id)
+            => _map.ContainsKey(id);
+
+        public bool TryGetState(DocumentId documentId, [NotNullWhen(true)] out TState? state)
+            => _map.TryGetValue(documentId, out state);
+
+        public TState? GetState(DocumentId documentId)
+            => _map.TryGetValue(documentId, out var state) ? state : null;
+
+        public TState GetRequiredState(DocumentId documentId)
+            => _map.TryGetValue(documentId, out var state) ? state : throw ExceptionUtilities.Unreachable;
+
+        /// <summary>
+        /// <see cref="DocumentId"/>s in the order in which they were added to the project (the compilation order).
+        /// </summary>
+        public readonly IReadOnlyList<DocumentId> Ids => _ids;
+
+        /// <summary>
+        /// States ordered by <see cref="DocumentId"/>.
+        /// </summary>
+        public IEnumerable<TState> States
+            => _map.Values;
+
+        /// <summary>
+        /// Get states ordered in compilation order.
+        /// </summary>
+        /// <returns></returns>
+        public IEnumerable<TState> GetStatesInCompilationOrder()
+        {
+            foreach (var id in Ids)
+            {
+                yield return _map[id];
+            }
+        }
+
+        public ImmutableArray<TValue> SelectAsArray<TValue>(Func<TState, TValue> selector)
+        {
+            using var _ = ArrayBuilder<TValue>.GetInstance(out var builder);
+
+            foreach (var (_, state) in _map)
+            {
+                builder.Add(selector(state));
+            }
+
+            return builder.ToImmutable();
+        }
+
+        public ImmutableArray<TValue> SelectAsArray<TValue, TArg>(Func<TState, TArg, TValue> selector, TArg arg)
+        {
+            using var _ = ArrayBuilder<TValue>.GetInstance(out var builder);
+
+            foreach (var (_, state) in _map)
+            {
+                builder.Add(selector(state, arg));
+            }
+
+            return builder.ToImmutable();
+        }
+
+        public async ValueTask<ImmutableArray<TValue>> SelectAsArrayAsync<TValue, TArg>(Func<TState, TArg, CancellationToken, ValueTask<TValue>> selector, TArg arg, CancellationToken cancellationToken)
+        {
+            using var _ = ArrayBuilder<TValue>.GetInstance(out var builder);
+
+            foreach (var (_, state) in _map)
+            {
+                builder.Add(await selector(state, arg, cancellationToken).ConfigureAwait(true));
+            }
+
+            return builder.ToImmutable();
+        }
+
+        public TextDocumentStates<TState> AddRange(ImmutableArray<TState> states)
+            => new(_ids.AddRange(states.Select(state => state.Id)),
```

Comment:
similar to above, just wondering if duplicates are supported or if this should throw

Reply:
The constructor calls `ToImmutableSortedDictionary`, which throws if there are duplicates.